Gazelle inconsistent proto dep referencing

Firstly, the proto support and the fact that gogo is abstracted is awesome ! thanks !

I working on a build for [lyft/protoc-gen-star](
https://github.com/hsyed/protoc-gen-star/tree/bazel_build) proto code generation framework.  Please see comment in snippet below  ([original file](https://github.com/hsyed/protoc-gen-star/blob/bazel_build/BUILD.bazel)).

```py
go_library(
    name = "go_default_library",
    srcs = [...],
    importpath = "github.com/lyft/protoc-gen-star",
    visibility = ["//visibility:public"],
    # Keep referencing the plugin and descriptor libraries directly. using the WKT's declared in `io_bazel_rules_go`
    # causes problems.
    # gazelle will prefer these by default.
    #   "@io_bazel_rules_go//proto/wkt:compiler_plugin_go_proto",
    #   "@io_bazel_rules_go//proto/wkt:descriptor_go_proto",
    #keep
    deps = [
        "@com_github_golang_protobuf//descriptor:go_default_library",
        "@com_github_golang_protobuf//proto:go_default_library",
        "@com_github_golang_protobuf//protoc-gen-go/descriptor:go_default_library",
        "@com_github_golang_protobuf//protoc-gen-go/generator:go_default_library",
        "@com_github_golang_protobuf//protoc-gen-go/plugin:go_default_library",
        "@com_github_spf13_afero//:go_default_library",
        "@org_golang_x_sync//errgroup:go_default_library",
    ],
)
```


if I remove the `#keep` from the library I get a warning (tests still pass):
```
GoLink: warning: package "github.com/golang/protobuf/protoc-gen-go/descriptor" is provided by more than one rule:
    @io_bazel_rules_go//proto/wkt:descriptor_go_proto
    @com_github_golang_protobuf//protoc-gen-go/descriptor:go_default_library
```


 if I remove the `#keep` from the tests I get a failure:

```
➜  protoc-gen-star git:(bazel_dep) ✗ bazel test all_tests
INFO: Analysed target //:go_default_test (1 packages loaded).
INFO: Found 1 test target...
ERROR: /Volumes/work/src/github.com/lyft/protoc-gen-star/BUILD.bazel:72:1: GoCompile darwin_amd64_stripped/go_default_test~/github.com/lyft/protoc-gen-star.a failed (Exit 1)
external/com_github_golang_protobuf/protoc-gen-go/descriptor/descriptor.pb.go:1265:0: inconsistent definition for type descriptor.FileOptions during import
        struct { JavaPackage *string "protobuf:\"bytes,1,opt,name=java_package,json=javaPackage\" json:\"java_package,omitempty\""; JavaOuterClassname *string "protobuf:\"bytes,8,opt,name=java_outer_classname,json=javaOuterClassname\" json:\"java_outer_classname,omitempty\""; JavaMultipleFiles *bool "protobuf:\"varint,10,opt,name=java_multiple_files,json=javaMultipleFiles,def=0\" json:\"java_multiple_files,omitempty\""; JavaGenerateEqualsAndHash *bool "protobuf:\"varint,20,opt,name=java_generate_equals_and_hash,json=javaGenerateEqualsAndHash\" json:\"java_generate_equals_and_hash,omitempty\""; JavaStringCheckUtf8 *bool "protobuf:\"varint,27,opt,name=java_string_check_utf8,json=javaStringCheckUtf8,def=0\" json:\"java_string_check_utf8,omitempty\""; OptimizeFor *descriptor.FileOptions_OptimizeMode "protobuf:\"varint,9,opt,name=optimize_for,json=optimizeFor,enum=google.protobuf.FileOptions_OptimizeMode,def=1\" json:\"optimize_for,omitempty\""; GoPackage *string "protobuf:\"bytes,11,opt,name=go_package,json=goPackage\" json:\"go_package,omitempty\""; CcGenericServices *bool "protobuf:\"varint,16,opt,name=cc_generic_services,json=ccGenericServices,def=0\" json:\"cc_generic_services,omitempty\""; JavaGenericServices *bool "protobuf:\"varint,17,opt,name=java_generic_services,json=javaGenericServices,def=0\" json:\"java_generic_services,omitempty\""; PyGenericServices *bool "protobuf:\"varint,18,opt,name=py_generic_services,json=pyGenericServices,def=0\" json:\"py_generic_services,omitempty\""; PhpGenericServices *bool "protobuf:\"varint,42,opt,name=php_generic_services,json=phpGenericServices,def=0\" json:\"php_generic_services,omitempty\""; Deprecated *bool "protobuf:\"varint,23,opt,name=deprecated,def=0\" json:\"deprecated,omitempty\""; CcEnableArenas *bool "protobuf:\"varint,31,opt,name=cc_enable_arenas,json=ccEnableArenas,def=0\" json:\"cc_enable_arenas,omitempty\""; ObjcClassPrefix *string "protobuf:\"bytes,36,opt,name=objc_class_prefix,json=objcClassPrefix\" json:\"objc_class_prefix,omitempty\""; CsharpNamespace *string "protobuf:\"bytes,37,opt,name=csharp_namespace,json=csharpNamespace\" json:\"csharp_namespace,omitempty\""; SwiftPrefix *string "protobuf:\"bytes,39,opt,name=swift_prefix,json=swiftPrefix\" json:\"swift_prefix,omitempty\""; PhpClassPrefix *string "protobuf:\"bytes,40,opt,name=php_class_prefix,json=phpClassPrefix\" json:\"php_class_prefix,omitempty\""; PhpNamespace *string "protobuf:\"bytes,41,opt,name=php_namespace,json=phpNamespace\" json:\"php_namespace,omitempty\""; PhpMetadataNamespace *string "protobuf:\"bytes,44,opt,name=php_metadata_namespace,json=phpMetadataNamespace\" json:\"php_metadata_namespace,omitempty\""; RubyPackage *string "protobuf:\"bytes,45,opt,name=ruby_package,json=rubyPackage\" json:\"ruby_package,omitempty\""; UninterpretedOption []*descriptor.UninterpretedOption "protobuf:\"bytes,999,rep,name=uninterpreted_option,json=uninterpretedOption\" json:\"uninterpreted_option,omitempty\""; XXX_NoUnkeyedLiteral struct {} "json:\"-\""; proto.XXX_InternalExtensions "json:\"-\""; XXX_unrecognized []byte "json:\"-\""; XXX_sizecache int32 "json:\"-\"" } (in "github.com/golang/protobuf/protoc-gen-go/plugin")
        struct { JavaPackage *string "protobuf:\"bytes,1,opt,name=java_package,json=javaPackage\" json:\"java_package,omitempty\""; JavaOuterClassname *string "protobuf:\"bytes,8,opt,name=java_outer_classname,json=javaOuterClassname\" json:\"java_outer_classname,omitempty\""; JavaMultipleFiles *bool "protobuf:\"varint,10,opt,name=java_multiple_files,json=javaMultipleFiles,def=0\" json:\"java_multiple_files,omitempty\""; JavaGenerateEqualsAndHash *bool "protobuf:\"varint,20,opt,name=java_generate_equals_and_hash,json=javaGenerateEqualsAndHash\" json:\"java_generate_equals_and_hash,omitempty\""; JavaStringCheckUtf8 *bool "protobuf:\"varint,27,opt,name=java_string_check_utf8,json=javaStringCheckUtf8,def=0\" json:\"java_string_check_utf8,omitempty\""; OptimizeFor *descriptor.FileOptions_OptimizeMode "protobuf:\"varint,9,opt,name=optimize_for,json=optimizeFor,enum=google.protobuf.FileOptions_OptimizeMode,def=1\" json:\"optimize_for,omitempty\""; GoPackage *string "protobuf:\"bytes,11,opt,name=go_package,json=goPackage\" json:\"go_package,omitempty\""; CcGenericServices *bool "protobuf:\"varint,16,opt,name=cc_generic_services,json=ccGenericServices,def=0\" json:\"cc_generic_services,omitempty\""; JavaGenericServices *bool "protobuf:\"varint,17,opt,name=java_generic_services,json=javaGenericServices,def=0\" json:\"java_generic_services,omitempty\""; PyGenericServices *bool "protobuf:\"varint,18,opt,name=py_generic_services,json=pyGenericServices,def=0\" json:\"py_generic_services,omitempty\""; PhpGenericServices *bool "protobuf:\"varint,42,opt,name=php_generic_services,json=phpGenericServices,def=0\" json:\"php_generic_services,omitempty\""; Deprecated *bool "protobuf:\"varint,23,opt,name=deprecated,def=0\" json:\"deprecated,omitempty\""; CcEnableArenas *bool "protobuf:\"varint,31,opt,name=cc_enable_arenas,json=ccEnableArenas,def=0\" json:\"cc_enable_arenas,omitempty\""; ObjcClassPrefix *string "protobuf:\"bytes,36,opt,name=objc_class_prefix,json=objcClassPrefix\" json:\"objc_class_prefix,omitempty\""; CsharpNamespace *string "protobuf:\"bytes,37,opt,name=csharp_namespace,json=csharpNamespace\" json:\"csharp_namespace,omitempty\""; SwiftPrefix *string "protobuf:\"bytes,39,opt,name=swift_prefix,json=swiftPrefix\" json:\"swift_prefix,omitempty\""; PhpClassPrefix *string "protobuf:\"bytes,40,opt,name=php_class_prefix,json=phpClassPrefix\" json:\"php_class_prefix,omitempty\""; PhpNamespace *string "protobuf:\"bytes,41,opt,name=php_namespace,json=phpNamespace\" json:\"php_namespace,omitempty\""; UninterpretedOption []*descriptor.UninterpretedOption "protobuf:\"bytes,999,rep,name=uninterpreted_option,json=uninterpretedOption\" json:\"uninterpreted_option,omitempty\""; XXX_NoUnkeyedLiteral struct {} "json:\"-\""; proto.XXX_InternalExtensions "json:\"-\""; XXX_unrecognized []byte "json:\"-\""; XXX_sizecache int32 "json:\"-\"" } (in "github.com/golang/protobuf/protoc-gen-go/generator")
GoCompile: error running subcommand: exit status 2
Target //:go_default_test failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 0.790s, Critical Path: 0.25s
INFO: 0 processes.
FAILED: Build did NOT complete successfully
```

--- a/README.rst
+++ b/README.rst
@@ -31,6 +31,7 @@ Go rules for Bazel_
 .. _Travis configuration file: .travis.yml
 .. _rules_go and Gazelle roadmap: roadmap.rst
 .. _Deprecation schedule: deprecation.rst
+.. _Avoiding conflicts: proto/core.rst#avoiding-conflicts
 
 .. ;; And now we continue with the actual content
 
@@ -525,4 +526,7 @@ How do I get information about the Go SDK used by rules_go?
 You can run: ``bazel build @io_bazel_rules_go//:go_info`` which outputs
 ``go_info_report`` with information like the used Golang version.
 
+How do I avoid conflicts with protocol buffers?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+See `Avoiding conflicts`_ in the proto documentation.

--- a/proto/core.rst
+++ b/proto/core.rst
@@ -13,6 +13,7 @@ Go Protocol buffers
 .. _Bourne shell tokenization: https://docs.bazel.build/versions/master/be/common-definitions.html#sh-tokenization
 .. _gogoprotobuf: https://github.com/gogo/protobuf
 .. _compiler.bzl: compiler.bzl
+.. _bazelbuild/bazel#3867: https://github.com/bazelbuild/bazel/issues/3867
 
 .. role:: param(kbd)
 .. role:: type(emphasis)
@@ -55,6 +56,99 @@ protoc), you can implement a compatible rule that returns one of these.
 The ``go_proto_library`` rule produces the normal set of `Go providers`_. This
 makes it compatible with other Go rules for use in ``deps`` and ``embed``
 attributes.
+
+Avoiding conflicts
+------------------
+
+When linking programs that depend on protos, care must be taken to ensure that
+the same proto isn't registered by more than one package. This may happen if
+you depend on a ``go_proto_library`` and a vendored ``go_library`` generated
+from the same .proto files. You may see compile-time, link-time, or run-time
+errors as a result of this.
+
+There are two main ways to avoid conflicts.
+
+Option 1: Use go_proto_library exclusively
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can avoid proto conflicts by using ``go_proto_library`` to generate code
+at build time and avoiding ``go_library`` rules based on pre-generated .pb.go
+files.
+
+Gazelle generates rules in this mode by default. When .proto files are present,
+it will generate ``go_proto_library`` rules and ``go_library`` rules that embed
+them (which are safe to use). Gazelle will automatically exclude .pb.go files
+that correspond to .proto files. If you have .proto files belonging to multiple
+packages in the same directory, add the following directives to your
+root build file:
+
+.. code:: bzl
+
+    # gazelle:proto package
+    # gazelle:proto_group go_package
+
+rules_go provides ``go_proto_library`` rules for commonly used proto libraries.
+The Well Known Types can be found in the ``@io_bazel_rules_go//proto/wkt``
+package. There are implicit dependencies of ``go_proto_library`` rules
+that use the default compiler, so they don't need to be written
+explicitly in ``deps``. You can also find rules for Google APIs and gRPC in
+``@go_googleapis//``. You can list these rules with the commands:
+
+.. code:: bash
+
+    $ bazel query 'kind(go_proto_library, @io_bazel_rules_go//proto/wkt:all)'
+    $ bazel query 'kind(go_proto_library, @go_googleapis//...)'
+
+Some commonly used Go libraries, such as ``github.com/golang/protobuf/ptypes``,
+depend on the Well Known Types. In order to avoid conflicts when using these
+libraries, separate versions of these libraries are provided with
+``go_proto_library`` dependencies. Gazelle resolves imports of these libraries
+automatically. For example, it will resolve ``ptypes`` as
+``@com_github_golang_protobuf//ptypes:go_default_library_gen``.
+
+Option 2: Use pre-generated .pb.go files
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can also avoid conflicts by generating .pb.go files ahead of time and using
+those exclusively instead of using ``go_proto_library``. This may be a better
+option for established Go projects that also need to build with ``go build``.
+
+Gazelle can generate rules for projects built in this mode. Add the following
+comment to your root build file:
+
+.. code:: bzl
+
+    # gazelle:proto disable_global
+
+This prevents Gazelle from generating ``go_proto_library`` rules. .pb.go files
+won't be excluded, and all special cases for imports (such as ``ptypes``) are
+disabled.
+
+If you have ``go_repository`` rules in your ``WORKSPACE`` file that may
+have protos, you'll also need to add
+``build_file_proto_mode = "disable_global"`` to those as well.
+
+.. code:: bzl
+
+    go_repository(
+        name = "com_example_some_project",
+        importpath = "example.com/some/project",
+        tag = "v0.1.2",
+        build_file_proto_mode = "disable_global",
+    )
+
+A note on vendored .proto files
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Bazel can only handle imports in .proto files that are relative to a repository
+root directory. This means, for example, if you import ``"foo/bar/baz.proto"``,
+that file must be in the directory ``foo/bar``, not
+``vendor/example.com/repo/foo/bar``.
+
+If you have proto files that don't conform to this convention, follow the
+instructions for using pre-generated .pb.go files above.
+
+The Bazel tracking issue for supporting this is `bazelbuild/bazel#3867`_.
 
 API
 ---

